breaking: modernize project

christian-bromann opened the pull request:
This patch will modernize the Ionicons project, it includes the following changes:

- [x] updates dependencies
- [x] update build scripts
- [x] ~update docs~ deferred to #1448

thetaPC commented on scripts/constants.ts:
```diff
@@ -0,0 +1,52 @@
+  'implements',
+  'instanceof',
+  'constructor',
+]);
```
Missing new line

```suggestion
]);

```

thetaPC commented on src/components/icon/request.ts:
```diff
@@ -5,53 +5,81 @@ const requests = new Map<string, Promise<any>>();
+   */
+  requests.set(url, req);
+  return req;
+}
```
Missing new line

```suggestion
}

```

thetaPC commented on scripts/readme.md:
The contributing link still points to the `master` branch. Let's update it to be `main`.

thetaPC commented on package.json:
```diff
@@ -13,34 +13,52 @@
-    "compile.scripts": "tsc -p scripts/tsconfig.json",
-    "build": "npm run compile.scripts && npm run build.files && npm run build.component && npm run collection.copy",
-    "build.files": "node scripts/build.js",
+    "build": "run-s clean build.*",
```
I'm getting an error when running this command:

```bash
> ionicons@7.3.1 build
> run-s clean build.*


> ionicons@7.3.1 clean
> rimraf dist components icons www

sh: rimraf: command not found
ERROR: "clean" exited with 127.
```

christian-bromann replied:
I've defined `rimraf` as dev dependency. Mind running `npm install` and try again?

thetaPC reviewed (changes requested):
The cheat sheet has broken links when clicking on an icon, might be worth fixing.

It might beneficial to add more comments of what's going on. We don't often work in this repo so comments would be a nice refresher.

christian-bromann commented:
@thetaPC thanks for reviewing your changes. I addressed the issues:

- added prettier commands to `package.json`
- added a copy output target to the `www` output target to properly copy over the SVGs
- updated `CONTRIBUTING.md` and CoC files
- added comments to `icon.tsx`

Let me know if you have further feedback!

thetaPC commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,151 @@
+   - Include relevant tests
+   - Make sure all tests pass: `npm test`
+   - Update documentation if needed
+   - [Create a new pull request](https://github.com/ionic-team/ionicons/compare) with the `main` branch as the `
```
Seems like this sentence got cut off:

```suggestion
   - [Create a new pull request](https://github.com/ionic-team/ionicons/compare) with the `main` branch as the base.
```

thetaPC commented on package.json:
Can we get rid of `build.component`? It seems redundant.

christian-bromann replied:
How so? It seems it is the command that calls `stencil build`. Or are you referring to `build.collection`?

thetaPC replied:
`stencil build` is already being called through `start.stencil`. I guess my real question is when should we be using `build.component`?

christian-bromann replied:
We use it indirectly when we build the project in [our pipeline](https://github.com/christian-bromann/ionicons/blob/f99664b2859a1277668cbb98c6474e0d0c8a2c49/.github/workflows/actions/build-core/action.yml#L22-L24).

thetaPC replied:
Works for me!

thetaPC commented on package.json:
```diff
@@ -13,34 +13,59 @@
-    "start": "npm run compile.scripts && npm run build.files && stencil build --dev --watch --serve",
-    "test": "npm run test.spec",
-    "test.spec": "stencil test --spec"
+    "build.collection": "tsx scripts/collection-copy.js",
```
Can we get rid of `build.collection`? It doesn't seem to be used anywhere. If we still need it then should it be?:

```suggestion
    "build.collection": "tsx scripts/collection-copy.ts",
```

If the command is needed, then let's add a comment on `collection-copy` explaining what the purpose of the script is. It's not clear to me what it's for.

thetaPC commented on package.json:
```diff
@@ -13,34 +13,59 @@
-    "build": "npm run compile.scripts && npm run build.files && npm run build.component && npm run collection.copy",
-    "build.files": "node scripts/build.js",
+    "build": "run-s clean build.* prettier",
+    "build.files": "tsx scripts/build.js",
```
Should this have the `ts` extension instead?

```suggestion
    "build.files": "tsx scripts/build.ts",
```

thetaPC commented on src/components/icon/test/icon.e2e.ts:
```diff
@@ -30,7 +30,11 @@ test.describe('icon: basic', () => {
     const rtlTests = page.locator('#rtl-tests');
-    await expect(rtlTests).toHaveScreenshot(`icon-rtl-diff.png`);
+    await expect(rtlTests).toHaveScreenshot(`icon-rtl-diff.png`, {
+      // Add more tolerance for font rendering differences
+      maxDiffPixelRatio: 0.02, // Allow up to 2% difference (much higher than the 0.01 difference in your error)
+      threshold: 0.2, // Increase threshold for per-pixel comparison
```
Is this still needed when the Playwright config has been [updated](https://github.com/ionic-team/ionicons/pull/1447/files#diff-f679bf1e58e8dddfc6cff0fa37c8e755c8d2cfc9e6b5dc5520a5800beba59a92R61-R71)?

christian-bromann replied:
good catch, fixed it.

brandyscarney commented on package.json:
```diff
@@ -13,34 +13,59 @@
+    "prettier": "npm run prettier.base -- --write",
+    "prettier.base": "prettier --cache \"./({bin,scripts,src,test}/**/*.{ts,tsx,js,jsx})|bin/stencil|.github/(**/)?*.(yml|yaml)|*.js\"",
+    "prettier.dry-run": "npm run prettier.base -- --list-different",
+    "start": "run-s build.files start.stencil",
```
I am getting this error on `npm run start`:

```bash
npm start


> ionicons@7.4.0 start
> run-s build.files start.stencil


> ionicons@7.4.0 build.files
> tsx scripts/build.js

[Error: ENOENT: no such file or directory, open '/Users/brandyscarney/Developer/ionicons/dist/ionicons/svg/accessibility.svg'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/brandyscarney/Developer/ionicons/dist/ionicons/svg/accessibility.svg'
}
ERROR: "build.files" exited with 1.
```

I ran `npm i && npm run build` prior to this. 

I followed the steps in the [contributing guide](https://github.com/ionic-team/ionicons/blob/a97e71239e4fa020a9a616abdd631319b3ae815c/CONTRIBUTING.md#environment-setup).

thetaPC replied:
Does it still happen after running `npm install` again? This was happening to me yesterday but ran fine after the second attempt.

christian-bromann replied:
It seems to fail every second time you call it, I will fix that.

brandyscarney commented on src/components/icon/request.ts:
```diff
@@ -5,53 +5,86 @@ const requests = new Map<string, Promise<any>>();
-    if (typeof fetch !== 'undefined' && typeof document !== 'undefined') {
+export const getSvgContent = (url: string, sanitize: boolean): Promise<string> => {
+  /**
+   * see if we already have a request for this url
```
```suggestion
   * See if we already have a request for this url
```

brandyscarney commented on src/components/icon/request.ts:
```diff
@@ -5,53 +5,86 @@ const requests = new Map<string, Promise<any>>();
+    .catch(() => safeFallback(url));
+
+  /**
+   * cache for the same requests
```
```suggestion
   * Cache for the same requests
```

christian-bromann commented on package.json:
```diff
@@ -11,36 +11,58 @@
+    "start.stencil": "stencil build --dev --watch --serve",
+    "test": "run-s test.spec",
+    "test.spec": "stencil test --spec",
+    "test.e2e": "playwright test",
```
it works on my machine™

<img width="634" alt="Screenshot 2025-04-29 at 10 17 53 AM" src="https://github.com/user-attachments/assets/a38d8359-f531-4b82-aa93-cd10868fe478" />

brandyscarney replied:
Sorry it's my first time using Stencil and I stopped the server to run the e2e tests. 😂

brandyscarney commented on src/svg/bowling-ball-outline.svg:
Do you know why `fill="currentColor"` was added to these 3 svgs? I don't see it in any others except for `src/svg/chevron-collapse-sharp.svg`

brandyscarney replied:
![Screenshot 2025-04-29 at 1 22 38 PM](https://github.com/user-attachments/assets/699911aa-c912-4c00-8ed1-cba1e2aa2ec8)

I tested setting `color` on each of them without `fill="currentColor"` and they seem to still work fwiw

brandyscarney replied:
Sorry I didn't see your reply when I added my 2nd comment. We ran into some issues in the past with `stroke:#000` being added to the svgs and it caused issues with being able to override the color, so I am always overly cautious here, but this change shouldn't do that.

I am curious how to reproduce the problem in https://github.com/ionic-team/ionicons/issues/1420 though because I tried setting the color directly and still see it inheriting on the dots:

```html
<ion-icon style="color: red" name="bowling-ball-outline"></ion-icon>
<ion-icon style="color: red" name="chevron-collapse-sharp"></ion-icon>
<ion-icon style="color: red" name="information-circle-outline"></ion-icon>
<ion-icon style="color: red" name="warning-outline"></ion-icon>

<ion-icon color="primary" name="bowling-ball-outline"></ion-icon>
<ion-icon color="secondary" name="chevron-collapse-sharp"></ion-icon>
<ion-icon color="tertiary" name="information-circle-outline"></ion-icon>
<ion-icon color="success" name="warning-outline"></ion-icon>
```

![Screenshot 2025-04-29 at 1 37 56 PM](https://github.com/user-attachments/assets/412f58fe-a991-4173-a390-31d4cef99b2b)

brandyscarney replied:
Here's a related issue: https://github.com/ionic-team/ionicons/issues/1129

brandyscarney reviewed (approved):
Everything looks good in my testing! Non-blocking comment on the svgs adding `fill="currentColor"`. 

It might be helpful if we add a note in the Contributing guide on how to setup and run the Playwright tests (including installing Playwright), but I don't think it's required for this PR.

Thanks for doing this! Great work. 🚀 

One thing to note — we should discuss whether or not this should be merged into `main` or if we should put it into a feature branch.

thetaPC reviewed (approved):
LGTM, I only have a non-blocking comment.

christian-bromann commented:
> we should discuss whether or not this should be merged into `main` or if we should put it into a feature branch.

Can you elaborate on this?

brandyscarney commented:
As discussed offline, this will be merged into `main` and released as a major.
